### ISSUE TITLE
[css-overflow-5]: mechanism to override and set semantic props on overflow markers and buttons

### ISSUE THREAD

**Issue by smhigley** [w3c/csswg-drafts#12269]
Some potential semantics that authors might need to add include (but are not limited to, there are probably weird edge cases I'm not thinking of):
- `role`
- `disabled` / `aria-disabled`
- `aria-labelledby`
- `aria-describedby` / `aria-description`

Using `aria-labelledby` could be seen as unnecessary because of the `content` approach, but that may not work for everyone -- often enough devs may have access to an element that should label the tab but not to the content of that element (e.g. when creating a design system, or with user-generated content). In those cases, it should be possible to name the scroll markers via reference rather than with a direct string.

The need for a description is probably a bit more self-explanatory, and isn't currently covered by CSS. Role as well -- I'm entirely sure people will use this in a way that doesn't cleanly fit the heuristics used to apply `link` or `tab`, and some accessibility SME who comes in later to fix it will need to override the semantics 😅.

One example of the need for a `disabled` state could be a multi-step form or process, where it's possible to navigate back through previous steps, but future steps are disabled. This would currently possible to implement for sighted users with greyed-out styles, but would need disabled semantics to be accessible for screen reader users.

**Comment by smhigley** [comment]
tagging @flackr because he made the mistake of talking to me about css overflow a few times

**Comment by flackr** [comment]
I think in the cases where authors need precise control of the semantic props they will currently need to use full elements with the scroll command (in place of `::scroll-button`) or links with the `scroll-target-group` property (in place of `::scroll-marker`). That said, I'd be happy to explore how we can improve things for the pseudo-elements where possible. Would you be willing to write some code snippets to exemplify these cases and what we want to be able to do? I'm having a hard time conceptually figuring out whether the labeled by example should use a `::scroll-marker`.

**Comment by smhigley** [comment]
Hm, I kind of view the response of "well if you want to make a good experience for screen reader users, you need to use the other approach" with some concern for a few reasons:
- The people doing accessibility work are often pulled in later in the process after the DOM is already stable. This approach effectively makes remediation impossible in many cases.
- A carousel pattern has a high degree of plasticity in how it is used, so it has highly variable case-by-case semantics (as can be seen in the [Chrome carousel gallery](https://chrome.dev/carousel/), where I can easily see use cases for tabs, non-interactive lists, links, and presentational scroll markers). Changing the role, focusability, method of defining the accName, and accDescription is, in my view, the 90% use case, not the 10%. If CSS scroll markers can't provide that, then it seems like a footgun to ship this authoring method at all.
- CSS content is set up poorly for localization in several ways, so the dev best practice would be to use `aria-labelledby` pointing to visible text:
    -  I've been seeing more AI-based string extraction & translation tools, and I have a hard time seeing how they would reliably manage extracting translatable strings from custom data attributes (while avoiding the custom data attributes with strings that should not be translated).
    -  CSS content doesn't currently get translated by user-facing translation tools, and it seems likely it won't in the foreseeable future either.
    - It's not possible to set lang/dir on CSS content, as far as I'm aware. If a scroll marker needs to set a custom lang/dir on its content, I don't really see how that could be achieved.

I'm guessing the reason this hasn't come up yet as a point of frustration is that these issues are the sort that tend to show up while doing QA and testing rather than during initial development.

**Comment by smhigley** [comment]
There's also one other issue I have yet to see addressed -- how does one add accessible names (let alone localizable ones) for `::column` scroll markers? The examples in the carousel gallery currently show them as unnamed.

**Comment by flackr** [comment]
> Hm, I kind of view the response of "well if you want to make a good experience for screen reader users, you need to use the other approach" with some concern for a few reasons:

I know when we discussed this before we considered the possibility of having an attribute on an element which gives a name to be used by content linking to that element (e.g. a scroll marker).

On the TAG review [cascading attribute sheets](https://gist.github.com/tabatkins/4074abaf486af5b3f7ac289737e216e4) were brought up as a possible future extension to allow specifying attributes of css pseudo-elements.

>   * I've been seeing more AI-based string extraction & translation tools, and I have a hard time seeing how they would reliably manage extracting translatable strings from custom data attributes (while avoiding the custom data attributes with strings that should not be translated).

What do these tools do? Do they try to modify strings on existing sites? Most sites I've seen using tooling to do translation have a way to provide translation from their framework and I suspect could set a custom attribute used by the marker to the string.

>   * CSS content doesn't currently get translated by user-facing translation tools, and it seems likely it won't in the foreseeable future either.

I'm not sure I understand this one, user-facing translation tools are capable of seeing the full document and could provide overrides of the accessible string in a stylesheet rule. Presumably this is an existing problem for other uses of the `content` property that would be great to work on better solutions to make it easier to do.

>   * It's not possible to set lang/dir on CSS content, as far as I'm aware. If a scroll marker needs to set a custom lang/dir on its content, I don't really see how that could be achieved.

You mentioned this on #12268 as well, should we discuss possible solutions here? there? in a separate issue?

> There's also one other issue I have yet to see addressed -- how does one add accessible names (let alone localizable ones) for ::column scroll markers? The examples in the carousel gallery currently show them as unnamed.

Authors can add accessible names the same way. I have an example on https://flackr.github.io/web-demos/css-overflow/scroll-marker-type/paged.html where I have labelled the page number:

```css
&::column::scroll-marker {
  counter-increment: --page;
  content: "" / "Page " counter(--page);
}
```

To make it more easily localizable using current capabilities I guess you would need to use an attribute instead of the "Page" string, e.g.

```css
  content: "" / attr(data-page) " " counter(--page);
```

And I imagine more generally where the number should show up may vary by language so you may need

```css
  content: "" / attr(data-page-before) " " counter(--page) " " attr(data-page-after);
```

Unless / until we have a way in CSS to perform substitution of parts of strings or parameter insertion.

**Comment by smhigley** [comment]
The string extraction tools I was looking at were something of an all-in-one solution to scan/catch hardcoded strings. I think right now there's a big unknown in how this works with the ecosystem of string management tools in general that comes down to the issue that CSS content is not currently a common place to put non-decorative text.

The issue with user-facing translation tools is that affecting CSS `content` strings would require writing a new stylesheet to the DOM, which is very different from modifying an attribute or content string in-place. I'm not a browser dev so I can't dive into specifics, but what I recall from conversations at TPAC was that there were some problems with doing so that appeared blocking.

On the column scroll marker accName issue -- while simple demos may get away with just a static string + counter, interactive controls in the wild need to be able to be given custom strings as their accName. Additionally, the string + number composition approach was something our component team actually researched some years ago for entirely different reasons, and we concluded it was too user-hostile and inflexible when integrating into existing localization workflows.
